src/planning/strategies/group-handle-connections.ts

```diff
@@ -24,83 +26,99 @@ export class GroupHandleConnections extends Strategy {
         if (recipe.getUnnamedUntypedConnections()) {
           return undefined;
         }
+        // All particles must have spec.
+        if (recipe.particles.some(p => !p.spec)) {
+          return undefined;
+        }
         // Find all unique types used in the recipe that have unbound handle connections.
         const types: Set<Type> = new Set();
-        recipe.getFreeConnections().forEach(hc => {
-          if (!Array.from(types).find(t => t.equals(hc.type))) {
-            types.add(hc.type);
+        recipe.getFreeConnections().forEach(({connSpec}) => {
+          if (!Array.from(types).find(t => t.equals(connSpec.type))) {
+            types.add(connSpec.type);
           }
         });
 
         const groupsByType = new Map();
-        types.forEach(type => {
-          // Find the particle with the largest number of unbound connections of the same type.
-          const countConnectionsByType = (connections: {[index: string]: HandleConnection}) => Object.values(connections).filter(conn => {
-            return !conn.isOptional && !conn.handle && type.equals(conn.type);
-          }).length;
+        for (const type of types) {
+          // Find the particle with the largest number of unbound required connections of the same type.
+          const countConnectionsByType = (particle: Particle) => {
+            return particle.spec.connections.filter(connSpec => {
+              return !connSpec.isOptional &&
+                     type.equals(connSpec.type) &&
+                     !particle.getConnectionByName(connSpec.name);
+            }).length;
```

Comment:
Move this extraction (unbound required connections of the same type) to a recipe helper function and make the type an optional parameter.

Reply:
moved to recipe/particle.ts